src/audio_core/sink/sink_stream.cpp

```diff
@@ -260,26 +247,31 @@ void SinkStream::ProcessAudioOutAndRender(std::span<s16> output_buffer, std::siz
     std::memcpy(&last_frame[0], &output_buffer[(frames_written - 1) * frame_size],
                 frame_size_bytes);
 
-    if (system.IsMulticore() && queued_buffers <= max_queue_size) {
-        Unstall();
+    {
+        std::scoped_lock lk{sample_count_lock};
+        last_sample_count_update_time =
+            Core::Timing::CyclesToUs(system.CoreTiming().GetClockTicks());
+        min_played_sample_count = max_played_sample_count;
+        max_played_sample_count += num_frames;
     }
 }
 
-void SinkStream::Stall() {
-    std::scoped_lock lk{stall_guard};
-    if (stalled_lock) {
-        return;
-    }
-    stalled_lock = system.StallApplication();
+u64 SinkStream::GetExpectedPlayedSampleCount() {
+    std::scoped_lock lk{sample_count_lock};
+    auto cur_time{Core::Timing::CyclesToUs(system.CoreTiming().GetClockTicks())};
+    auto time_delta{cur_time - last_sample_count_update_time};
+    auto exp_played_sample_count{min_played_sample_count +
+                                 (TargetSampleRate * time_delta) / std::chrono::seconds{1}};
+
+    return std::min<u64>(exp_played_sample_count, max_played_sample_count);
 }
 
-void SinkStream::Unstall() {
-    std::scoped_lock lk{stall_guard};
-    if (!stalled_lock) {
-        return;
+void SinkStream::WaitFreeSpace(std::stop_token stop_token) {
+    while (queued_buffers >= max_queue_size) {
+        std::unique_lock lk{release_mutex};
+        Common::CondvarWait(release_cv, lk, stop_token, [this]() { return release_signalled; });
+        release_signalled = false;
```

Comment:
Do you really need `release_signalled`? Couldn't you just wait for the underlying condition to be true, i.e.
```suggestion
    Common::CondvarWait(release_cv, lk, stop_token, [this]() { return queued_buffers < max_queue_size; });
```

And then in ProcessAudioOutAndRender, after dequeuing a buffer,
```cpp
            {
                std::unique_lock lk{release_mutex};
            }
            release_cv.notify_one();
```